wgpu-core/src/command/mod.rs

```diff
@@ -78,21 +78,185 @@ pub(crate) enum CommandEncoderStatus {
+    ///
+    /// Call [`Self::unlock_encoder`] to put the [`CommandBuffer`] back into the [`Self::Recording`] state.
+    fn lock_encoder(&mut self) -> Result<(), CommandEncoderError> {
+        match *self {
+            Self::Recording(_) => {
+                change_state_with_inner!(self, Self::Recording, Self::Locked);
+                Ok(())
+            }
+            Self::Finished(_) => Err(CommandEncoderError::NotRecording),
+            Self::Locked(_) => {
+                let _ = mem::replace(self, Self::Error);
+                Err(CommandEncoderError::Locked)
+            }
+            Self::Error => Err(CommandEncoderError::Invalid),
+        }
+    }
+
+    /// Unlocks the [`CommandBuffer`] and puts it back into the [`Self::Recording`] state.
+    ///
+    /// This function is the unlocking counterpart to [`Self::lock_encoder`].
+    ///
+    /// It is only valid to call this function if the encoder is in the [`Self::Locked`] state.
+    fn unlock_encoder(&mut self) -> Result<EncoderGuard<'_>, CommandEncoderError> {
+        match *self {
+            Self::Locked(_) => {
+                change_state_with_inner!(self, Self::Locked, Self::Recording);
+                Ok(EncoderGuard {
+                    inner: self,
+                    succeeded: false,
+                })
+            }
+            Self::Finished(_) => Err(CommandEncoderError::NotRecording),
+            Self::Recording(_) => {
+                let _ = mem::replace(self, Self::Error);
+                Err(CommandEncoderError::Invalid)
+            }
+            Self::Error => Err(CommandEncoderError::Invalid),
+        }
+    }
+
+    fn finish(&mut self, device: &Device) -> Result<(), CommandEncoderError> {
+        match self {
+            Self::Recording(inner) => {
+                if let Err(e) = inner.encoder.close(device) {
+                    let _ = mem::replace(self, Self::Error);
+                    Err(e.into())
+                } else {
+                    change_state_with_inner!(self, Self::Recording, Self::Finished);
+                    // Note: if we want to stop tracking the swapchain texture view,
+                    // this is the place to do it.
+                    Ok(())
+                }
+            }
+            Self::Finished(_) => Err(CommandEncoderError::NotRecording),
+            Self::Locked(_) => {
+                let _ = mem::replace(self, Self::Error);
+                Err(CommandEncoderError::Locked)
+            }
+            Self::Error => Err(CommandEncoderError::Invalid),
+        }
+    }
+}
+
+/// A guard on which [`EncoderGuard::succeeded`] should be called if
+/// no errors have been raised during encoding, otherwise the encoder will be put in the
+/// [`CommandEncoderStatus::Error`] on [`EncoderGuard::drop`].
+pub(crate) struct EncoderGuard<'a> {
+    inner: &'a mut CommandEncoderStatus,
+    succeeded: bool,
+}
+
+impl<'a> EncoderGuard<'a> {
+    pub(crate) fn succeeded(&mut self) {
```

Comment:
Would it make sense for this to take `self` by value?

Reply:
Oh, that's a great idea!